client/network/src/request_responses.rs

```diff
@@ -345,148 +352,185 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
-						}
+						// This `wake_by_ref` makes sure that `pending_responses` gets polled
+						// after we have added the new element.
+						cx.waker().wake_by_ref();
+					}
 
-						// Received a response from a remote to one of our requests.
-						RequestResponseEvent::Message {
-							message:
-								RequestResponseMessage::Response {
-									request_id,
-									response,
-								},
-							..
-						} => {
-							let out = Event::RequestFinished {
-								request_id,
-								result: response.map_err(|()| RequestFailure::Refused),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
-						}
+					// Received a response from a remote to one of our requests.
+					RequestResponseEvent::Message {
+						message:
+						RequestResponseMessage::Response {
+							request_id,
+							response,
+						},
+						..
+					} => {
+						let out = Event::RequestFinished {
+							request_id,
+							result: response.map_err(|()| RequestFailure::Refused),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// One of our requests has failed.
-						RequestResponseEvent::OutboundFailure {
+					// One of our requests has failed.
+					RequestResponseEvent::OutboundFailure {
+						request_id,
+						error,
+						..
+					} => {
+						let out = Event::RequestFinished {
 							request_id,
-							error,
-							..
-						} => {
-							let out = Event::RequestFinished {
-								request_id,
-								result: Err(RequestFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
-						}
+							result: Err(RequestFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// Remote has tried to send a request but failed.
-						RequestResponseEvent::InboundFailure { peer, error, .. } => {
-							let out = Event::InboundRequest {
-								peer,
-								protocol: protocol.clone(),
-								result: Err(ResponseFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					// Remote has tried to send a request but failed.
+					RequestResponseEvent::InboundFailure { request_id, peer, error, .. } => {
+						self.pending_responses_arrival_time.pop(&request_id);
+						let out = Event::InboundRequest {
+							peer,
+							protocol: protocol.clone(),
+							result: Err(ResponseFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
+					RequestResponseEvent::ResponseSent { request_id, peer } => {
+						match self.pending_responses_arrival_time.pop(&request_id) {
+							Some(arrival) => {
+								let out = Event::InboundRequest {
+									peer,
+									protocol: protocol.clone(),
+									result: Ok(arrival.elapsed()),
+								};
+								return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+							},
+							None => log::debug!(
```

Comment:
The `Event::InboundRequest` is used for statistic purposes only, thus, in my eyes, emitting the whole event on a best-effort basis, instead of making the duration optional, is fine.

With the current arrival time LRU cache size set to 1_000 I expect the possibility to ever hit this case to be low. In case we ever do, I would classify it as a bug, requiring the cache size to be raised or the whole tracking to be reworked.

As a note, `network/service.rs`, the consumer of this event, uses `Event::InboundRequest` only to record the duration, thus with a duration of `None` the event would be simply ignored.

With the above in mind, do you still think that one should emit an `Event::InboundRequest` with a duration `None`? If so, I am happy to do the change.

Reply:
> With the current arrival time LRU cache size set to 1_000 I expect the possibility to ever hit this case to be low. In case we ever do, I would classify it as a bug, requiring the cache size to be raised or the whole tracking to be reworked.

I guess therein lies the difference then, because I would not consider it a bug and would just settle on a limit that is sufficient to get enough data points for the metrics to be meaningful, which `1000` probably does well.

> As a note, network/service.rs, the consumer of this event, uses Event::InboundRequest only to record the duration, thus with a duration of None the event would be simply ignored.

I see. At the risk of sounding like a broken record and knowing that this ship probably sailed, I would personally likely not emit these events then in the first place and track the metrics right here. Nevertheless, keeping these events, I would still favor local reasoning here, meaning that the choice of what this code should do not be driven by knowing how the events will be consumed. So yes, as long as these events are emitted, I'd prefer emitting them without making assumptions about how they are consumed. This may just express my personal preference though, feel free to go with what you think is best.

Reply:
> At the risk of sounding like a broken record and knowing that this ship probably sailed, I would personally likely not emit these events then in the first place and track the metrics right here. 

I wouldn't go as far as saying that the ship has sailed, but I think it requires a separate discussion and preferably a single pull request changing all of `client/network` to this pattern.

> Nevertheless, keeping these events, I would still favor local reasoning here, meaning that the choice of what this code should do not be driven by knowing how the events will be consumed.

Convinced. Keeping complexity low by reasoning about this within request-responses.rs, thus providing a better abstraction sounds good to me. https://github.com/paritytech/substrate/pull/7696/commits/467206a4c801af6ad54adf93ec53158c9a3615b0 makes the required changes.